### PR TITLE
Push image to docker hub #651

### PR DESCRIPTION
**Pull request by AugustasV**
Edited github CI yaml file to upload artifact into dockerhub during build stage.

Issue #651

### CODE REVIEW

**Review by ytsarev — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution 👍 Curldemo is not connected to main k8gb release so it's better to create a separate flow for it. Also `env.GITHUB_JOB` will probably create unrelated tag name for the image

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -49,3 +49,11 @@ jobs:
...
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          context: ./deploy/test-apps/curldemo/
+          file: ./Dockerfile
+          push: true
+          tags: ${{ secrets.DOCKER_USER }}/k8gb:${{ env.GITHUB_JOB }}
```
Comment (ytsarev) [inline .github/workflows/release.yaml:59]: should be different docker repo as well.
```suggestion
          tags: ${{ secrets.DOCKER_USER }}/k8gb-demo:${{ env.GITHUB_JOB }}
```

**Review by ytsarev — CHANGES_REQUESTED** [review]

File: .github/workflows/curldemo.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+name: curldemo
+
+on:
+  push:
+    tags:
+      - "v*"
```
Comment (ytsarev) [inline .github/workflows/curldemo.yaml:6]: It's not connected to main release, so it is better to execute as reaction to file patch change, see https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#example-including-paths

File: .github/workflows/curldemo.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      DOCKER_CLI_EXPERIMENTAL: "enabled"
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Get tag
+        id: get_tag
+        run: |
+          previous_tag=$(git tag --sort=v:refname | tail -2 | head -1)
+          echo "::set-output name=previous_tag::${previous_tag}"
```
Comment (ytsarev) [inline .github/workflows/curldemo.yaml:21]: Should not be required here as it is not connected with the release

File: .github/workflows/curldemo.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      - name: Set up Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.16.9
+      - name: Login to Dockerhub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USER }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      - name: Run GoReleaser
```
Comment (ytsarev) [inline .github/workflows/curldemo.yaml:31]: no goreleaser required, we are not building any binary with curl demo

File: .github/workflows/release.yaml [hunk 5]
```diff
@@ -48,4 +48,4 @@ jobs:
           version: latest
           args: release --rm-dist --release-notes=changes
         env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (ytsarev) [inline .github/workflows/release.yaml:51]: please revert newline change

**Review by ytsarev — CHANGES_REQUESTED** [review]
Please address suggested changes.

Please also use --signoff to satisfy DCO https://github.com/k8gb-io/k8gb/pull/709/checks?check_run_id=4060681862 .

We are CNCF project and it is a strong requirement.

**Review by jkremser — CHANGES_REQUESTED** [review]

File: .github/workflows/curldemo.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          context: ./deploy/test-apps/curldemo/
+          file: ./Dockerfile
+          push: true
+          tags: ${{ secrets.DOCKER_USER }}/k8gb-demo:latest
```
[the remark this answers is not in the document]
Comment (AugustasV) [reply .github/workflows/curldemo.yaml]: Yes I know that using latest is evil, which env variable you would recommend to use? I tried right now with  GITHUB_RUN_NUMBER

[the remark this answers is not in the document]
Comment (ytsarev) [reply .github/workflows/curldemo.yaml]: @jkremser `latest` is evil but imho exactly what we need for the trivial demo. Obviously open to any better suggestions but the main goal is to keep it simple.

[the remark this answers is not in the document]
Comment (jkremser) [reply .github/workflows/curldemo.yaml]: make sense, kets keep it simple

File: .github/workflows/release.yaml (hunk 5 above)
Comment (ytsarev) [inline .github/workflows/release.yaml:51]: Please keep new line around

File: .github/workflows/curldemo.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+          username: ${{ secrets.DOCKER_USER }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+      - name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          context: ./deploy/test-apps/curldemo/
+          file: ./Dockerfile
+          push: true
+          tags: ${{ secrets.DOCKER_USER }}/k8gb-demo-curl:$GITHUB_RUN_NUMBER
```
Comment (ytsarev) [inline .github/workflows/curldemo.yaml:29]: Let's stick with `latest` as we agreed in https://github.com/k8gb-io/k8gb/pull/709#discussion_r740856251

**Review by ytsarev — CHANGES_REQUESTED** [review]

File: .github/workflows/curldemo.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+jobs:
+  curldemo:
+    runs-on: ubuntu-20.04
+    env:
+      DOCKER_CLI_EXPERIMENTAL: "enabled"
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
```
Comment (jkremser) [inline .github/workflows/curldemo.yaml:16]:
```suggestion
```

nit: `fetch-depth: 0` [indicates](https://github.com/actions/checkout/blob/ec3a7ce113134d7a93b817d10a8272cb61118579/README.md?plain=1#L92) that all history, branches and tags will be fetched, but we are not doing any fancy git stuff here, so the default shallow fetch should be ok. If omitted the `1` is used

**Review by jkremser — CHANGES_REQUESTED** [review]

**Review by jkremser — APPROVED** [review]

**Review by k0da — APPROVED** [review]
LGTM

**Review by ytsarev — APPROVED** [review]
@AugustasV looks great, thanks for all the fixes :)

**Pull request merged**